### PR TITLE
AE-1269 create discord channels for forta sentinels

### PR DESCRIPTION
**Pull request by darian-ae**
This PR is to create discord notifications for Forta Sentinels in the compound monitoring effort. This is done through autotask's, which will send notifications to our test Discord channel via a webhook when one of the compound forta bots creates an alert.

Note that the Distribution Monitor Bot is the only one with previous alerts. The rest of the Bot's have never picked up an alert (as they were only created early April), so alert data had to be mocked locally.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -29,6 +29,10 @@ The use of `describe` and `it` is currently only necessary because we are using
   "endBlockNumber": <integer_for_ending_block_number>
 }
 ```
+- Create a .env file that contains the name of your discord webook and the url for it:
```
Comment (arbitrary-echo) [inline README.md:32]:
```suggestion
- Create a .env file that contains the name of your Discord webhook and the url for it:
```

File: README.md [hunk 2]
```diff
@@ -29,6 +29,10 @@ The use of `describe` and `it` is currently only necessary because we are using
...
+  - ex.) `FortaSentinelTestingDiscord = "discord webhook url"`
+- The `cToken-monitor` and `governance` autotasks both have previous alerts, so the autotask event data can be fed through the handler by running the `autotask.spec.js` file. You can rename either of these files to `autotask.js` and run `npm test` to test it out.
+- For the `asset-monitor`, `large-borrows-governance` and `oracle-price-monitor` autotasks, these agents have not picked up on any previous findings and the alert to discord channels are invoked by passing in mock data.
```
Comment (arbitrary-echo) [inline README.md:35]:
```suggestion
- For the `asset-monitor`, `large-borrows-governance` and `oracle-price-monitor` Autotasks, these bots have not picked up on any previous findings and the alert to Discord channels are invoked by passing in mock data.
```

File: asset-monitor/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# Compound cToken Underlying Asset Monitor
+
+## Description
+
+This agent monitors the underlying asset of Compound Finance cToken contracts.  First
```
Comment (arbitrary-echo) [inline asset-monitor/README.md:5]:
```suggestion
This bot monitors the underlying asset of Compound Finance cToken contracts.  First
```

File: asset-monitor/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  - Severity is set to the pattern specific value in `agent-config.json`
+  - Metadata field contains:
+    - cToken symbol
+    - cToken address
+    - underlyingAssetAddress
+    - Arguments passed with event (e.g. for new implementation address)
+
+## Autotask
+
+This autotask will send alerts to the compound discord channel when the agent produces a finding.
```
Comment (arbitrary-echo) [inline asset-monitor/README.md:25]:
```suggestion
This Autotask will send alerts to the Compound Discord channel when the bot produces a finding.
```

File: README.md [hunk 5]
```diff
@@ -29,6 +29,10 @@ The use of `describe` and `it` is currently only necessary because we are using
   "endBlockNumber": <integer_for_ending_block_number>
 }
 ```
+- Create a .env file that contains the name of your discord webook and the url for it:
+  - ex.) `FortaSentinelTestingDiscord = "discord webhook url"`
+- The `cToken-monitor` and `governance` autotasks both have previous alerts, so the autotask event data can be fed through the handler by running the `autotask.spec.js` file. You can rename either of these files to `autotask.js` and run `npm test` to test it out.
```
Comment (arbitrary-echo) [inline README.md:34]:
```suggestion
- The `cToken-monitor` and `governance` Autotasks both have previous alerts, so the Autotask event data can be fed through the handler by running the `autotask.spec.js` file. You can rename either of these files to `autotask.js` and run `npm test` to test it out.
```

File: asset-monitor/package.json [hunk 6]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "forta-ae-comp-asset-monitor",
+  "version": "0.0.1",
+  "description": "This agent monitors the underlying asset of a Compound Finance cToken.",
```
Comment (arbitrary-echo) [inline asset-monitor/package.json:4]:
```suggestion
  "description": "This bot monitors the underlying asset of a Compound Finance cToken.",
```

File: asset-monitor/src/agent.spec.js [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            protocol: 'Compound',
+            findingType: 'INFORMATION',
+            hash: '0xcee8d4bd1c065260acdcfa51c955fc29c984145de2769b685f29701b6edf318f',
+            source: {
+              transactionHash: '0xaaec8f4fcb423b5190b8d78b9595376ca34aee8a50c7e3250b3a9e79688b734b',
+              block: {
+                number: 14496506,
+                chainId: 1,
+              },
+              agent: {
```
Comment (echoAE) [inline asset-monitor/src/agent.spec.js:45]:
```suggestion
              bot: {
```

File: oracle-price-monitor/src/autotask.js [hunk 8]
```diff
@@ -0,0 +1,270 @@
+// ORACLE PRICE ALERT - note there's never been an alert to this agent yet
+// agent id - 0x8a6e2ae3b279c561f7ba64e5011313df19e08b1f398d59d94dfbd4148e6cafce
+
+const axios = require('axios');
+const ethers = require('ethers');
+const needle = require('needle');
```
Comment (arbitrary-echo) [inline oracle-price-monitor/src/autotask.js:6]: Is `needle` available in the Defender environment? (I assume that it is, or the autotask would perpetually fail).

Comment (arbitrary-echo) [inline oracle-price-monitor/src/autotask.js:6]: Is `needle` available in the Defender environment? (I assume that it is, or the autotask would perpetually fail).

**Review by arbitrary-echo — COMMENTED** [review]
I've reviewed the files in this pull request, but I'm hesitant to merge this in here.  It may be best for us to close the current Jira ticket, but create another ticket to examine a better place for this testing work.

What is most important is that we have a way to test our Autotasks outside of the Defender environment, but because those Autotasks are dependent upon Forta Bots, we are using code that is already tracked somewhere else (so now we have extra copies of both Autotasks and Forta Bots).

Our goal should be one source of truth for our Autotasks and Forta Bots, but I'm uncertain of how to then integrate this testing code.  Perhaps the approach should be the following:
- Stop updating this separate repository
- Move the original code from this repository over to the `compound-monitoring` repository
- Write "shim code" that will automatically retrieve the backed up Autotask code and associated Forta Bot code
- In the test code, allow the default approach to be checking the Forta Public API for alerts, with the fallback of mocking the alert the way you have done here for some of the Bots

**Pull request closed without merging**